pkg/ruler/manager.go

```diff
@@ -176,6 +177,20 @@ func (r *DefaultMultiTenantManager) newManager(ctx context.Context, userID strin
 	return r.managerFactory(ctx, userID, notifier, r.logger, reg), nil
 }
 
+func (r *DefaultMultiTenantManager) deleteNotifier(userID string) {
+	r.notifiersMtx.Lock()
+	defer r.notifiersMtx.Unlock()
+
+	n, ok := r.notifiers[userID]
+	if !ok {
+		return
+	}
+
+	n.stop()
```

Comment:
I see `stop` is implemented as 

```
	rn.sdCancel()
	rn.notifier.Stop()
	rn.wg.Wait()
```

I wonder if stop may be long running, if so we should not call stop while holding lock.

Reply:
I can remove userID from `r.notifiers` first, so that even if we have concurrent cleanup, we won't have multiple calls to `.stop()` to the same notifier